tests/AsyncTest.php

```diff
@@ -84,4 +85,91 @@ public function testAsyncReturnsPromiseThatFulfillsWithValueWhenCallbackReturnsA
         $this->assertGreaterThan(0.1, $time);
         $this->assertLessThan(0.12, $time);
     }
+
+    public function testCancel()
+    {
+        self::expectOutputString('a');
+
+        $promise = async(function (): int {
+            echo 'a';
+            await(sleep(2));
+            echo 'b';
+
+            return time();
+        })();
+        Loop::addTimer(0.001, function () use ($promise) {
+            $promise->cancel();
+        });
+
+        Loop::run();
+    }
+
+    public function testNestedCancel()
+    {
+        self::expectOutputString('abc');
+
+        $promise = async(function (): int {
+            echo 'a';
+            await(async(function(): void {
+                echo 'b';
+                await(async(function(): void {
+                    echo 'c';
+                    await(sleep(2));
+                    echo 'd';
+                })());
+                echo 'e';
+            })());
+            echo 'f';
+
+            return time();
+        })();
+        Loop::addTimer(0.001, function () use ($promise) {
+            $promise->cancel();
+        });
+
+        Loop::run();
+    }
+
+    public function testCancelFiberThatCatchesExceptions()
+    {
+        self::expectOutputString('ab');
+        $this->expectException(\Exception::class);
+        $this->expectExceptionMessage('Timer cancelled');
+
+        $promise = async(function (): int {
+            echo 'a';
+            try {
+                await(sleep(2));
+            } catch (\Throwable) {
+                // No-Op
+            }
+            echo 'b';
+            await(sleep(0.1));
+            echo 'c';
+
+            return time();
+        })();
+
+        Loop::addTimer(0.001, function () use ($promise) {
+            $promise->cancel();
+        });
+        await($promise);
+    }
```

Comment:
@kelunik Added a test and the changes for the situation you mentioned on Twitter